tests/net/socket/socketpair/src/test_socketpair_block.c

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright (c) 2020 Friedt Professional Engineering Services, Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <fcntl.h>
+#include <string.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_test, CONFIG_NET_SOCKETS_LOG_LEVEL);
+
+#include <net/socket.h>
+#include <sys/util.h>
+#include <posix/unistd.h>
+
+#include <ztest_assert.h>
+
+#undef read
+#define read(fd, buf, len) zsock_recv(fd, buf, len, 0)
```

Comment:
Why are these defines? Why not use proper read()/write() functions?

Reply:
Because those don't seem to get mapped to the socket implementations and instead go to stdio :(

Reply:
> Because those don't seem to get mapped to the socket implementations and instead go to stdio :(

Well, there're definitely POSIX-complient write()/read() implementations, that's the whole idea of fdtable munging to support them. But POSIX is not first-class citizen in Zephyr, it's an optional subsystem, of which individual features should be enabled by config options. In this case, it's actually enabled by the master switch: https://github.com/zephyrproject-rtos/zephyr/blob/master/lib/os/fdtable.c#L159 .

But you definitely lead me to a "d'oh" moment: there're absolutely zero tests and samples using read/write with sockets. Well, what can I say, if you can in all seriousness receive a review like https://github.com/zephyrproject-rtos/zephyr/pull/24913#pullrequestreview-406830171 from project owners, not submitting more patches than strictly necessary can be a boon ;-).

And when patching a sample to see if it works, I hit https://github.com/zephyrproject-rtos/zephyr/issues/25083. Ah, a typical zephyr prerelease week.

Anyway, specific suggestion to deal with this: have something like:
~~~
#ifdef CONFIG_POSIX_API
#define READ(fd, buf, sz) read(fd, buf, sz) 
#else
#define READ(fd, buf, sz) zsock_recv(fd, buf, sz, 0) 
#endif 
~~~

Then have 2 test variants for both !CONFIG_POSIX_API and CONFIG_POSIX_API in test.yaml. Don't hesitate to ask for specific pointers if something's unclear. (Jokes aside, complexity of munging with multiple configurations is the reason of technical debt re: read/write.)

P.S. No hurries of course, but if this PR will be in good syntactic shape (commit message, no "stashing" commits, etc.), I'd be ready to +1 it tomorrow, as I think it's done well, and any rough edges can be addressed during the freeze window. (Obviously, that doesn't guarantee anything.) Thanks.

Reply:
I can remove the read / write if you like. But on connected sockets, read / write are synonymous with send / recv with 0 flags, and those in tern are synonymous with sendto / recvfrom with a NULL addr argument, and those are synonymous with sendmsg / recvmsg with... you probably see the pattern.

You can also call connect on UDP sockets as well. However, I think SOCK_DGRAM poses some unique challenges for socketpair though. It would be smarter to use the k_fifo for those, since it eliminates a lot of packet separation overhead. For that reason, I would save it to later *if* someone really wants it. Socketpair is normally used on streams, so not a priority.